assets/app/scripts/services/data.js

```diff
@@ -96,29 +97,41 @@ angular.module('openshiftConsole')
 
 // resource:  API resource (e.g. "pods")
 // context:   API context (e.g. {project: "..."})
-// callback:  function to be called with the list of the requested resource and context,
+// callback:  (optional) function to be called with the list of the requested resource and context,
 //            parameters passed to the callback:
 //            Data:   a Data object containing the (context-qualified) results
 //                    which includes a helper method for returning a map indexed
 //                    by attribute (e.g. data.by('metadata.name'))
-// opts:      options (currently none, placeholder)
+// opts:      errorNotification - will popup an error notification if the API request fails (default true)
+// returns promise that will be resolved or rejected when the list operation is complete
   DataService.prototype.list = function(resource, context, callback, opts) {
     resource = APIService.toResourceGroupVersion(resource);
     var callbacks = this._listCallbacks(resource, context);
-    callbacks.add(callback);
+    if (callback) {
+      callbacks.add(callback);
+    }
+
+    var deferred = $q.defer();
+    var promises = this._listPromises(resource, context);
+    promises.push(deferred);
 
```

Comment:
How do you feel about hiding away some of the details here to make the flow a little easier:
```javascript
  var key = this._uniqueKeyForResourceContext(resource, context);
  // currently we are inferring 'cached' from _watchInFlight and _resourceVersion
  if(this.cached(key)) {
    this.fireCallbacks(this._data(key));
    deferred.resolve(this._data(key));
  }
```